Bounds Refactor

adamperlin opened the pull request:
Closes #639 

_Briefly describe your proposed changes:_

Moves the task of relative time resolution to the planner. The planner's `BoundsSpec` struct now uses `values.Time` as opposed to `query.Time`. A new `query.Bounds` struct was added which uses `query.Time`. It is now used by flux functions instead of the `BoundsSpec`.

_What was the problem?_
Deferring time resolution past the planner meant that the current `now` context had to be passed around whenever bounds operations were performed while planning, which made the code unnecessarily messy.
_What was the solution?_
Resolve bounds directly in the planner, once for each set of bounds.

  - [ ] CHANGELOG.md updated with a link to the PR (not the Issue)
  - [x] Rebased/mergeable
  - [x] Tests pass
  - [ ] swagger.json updated (if modified Go structs or API)
  - [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

nathanielc commented on query/execute/executor.go:
```diff
@@ -128,13 +128,19 @@ func (es *executionState) createNode(ctx context.Context, pr *plan.Procedure, no
+	// Add explicit stream context if bounds are set on this node
+	var streamContext streamContext
+	if pr.Bounds != nil {
+		streamContext = newStreamContext(Bounds{
```
This code makes it read like the `streamContext` is a pointer. Instead why not just set `streamContext.Bound` to the correct bounds if the procedure bounds exist?
```
if pr.Bounds != nil {
    streamContext.Bounds = ...
}
```

nathanielc commented on query/plan/physical.go:
```diff
@@ -60,6 +67,24 @@ func NewPlanner() Planner {
+		return nil, errors.New("bounds contain zero time")
+	}
+	if bounds.IsEmpty(now) {
+		return nil, errors.New("bounds are empty")
```
Why are empty bounds an error?

adamperlin replied:
just changed per our conversation

nathanielc commented on query/plan/physical.go:
```diff
@@ -136,26 +161,31 @@ func (p *planner) Plan(lp *LogicalPlanSpec, s Storage) (*PlanSpec, error) {
 		// of the bounds of any parent procedure
 		pr.DoParents(func(parent *Procedure) {
-			if pr.Bounds.HasZero() {
+			if pr.Bounds != nil && parent.Bounds != nil {
```
I think this is simpler if you return early if `parent.Bounds == nil`. That way you only have to check it once.

nathanielc commented on query/plan/physical.go:
```diff
@@ -136,26 +161,31 @@ func (p *planner) Plan(lp *LogicalPlanSpec, s Storage) (*PlanSpec, error) {
-				if newBounds != EmptyBoundsSpec {
+			convertedBounds, err := ToBoundsSpec(bounded.TimeBounds(), pr.plan.Now)
+			if err != nil {
+				return nil, errors.Wrap(err, "invalid time bounds provided")
```
Can you add the context of which procedure caused the error?

nathanielc commented on query/plan/physical.go:
```diff
@@ -136,26 +161,31 @@ func (p *planner) Plan(lp *LogicalPlanSpec, s Storage) (*PlanSpec, error) {
+
+			if pr.Bounds != nil {
+				newBounds := pr.Bounds.Intersect(convertedBounds)
+				if !newBounds.IsEmpty() {
```
Why do we ignore the empty bounds?

nathanielc commented on query/plan/procedure.go:
```diff
@@ -105,88 +111,81 @@ type ProcedureKind string
-// are empty, i.e., if start >= stop.
-func (b BoundsSpec) IsEmpty(now time.Time) bool {
-	return b.Start.Time(now).Equal(b.Stop.Time(now)) || b.Start.Time(now).After(b.Stop.Time(now))
+func NewEmptyBoundsSpec() *BoundsSpec {
```
Why is this function useful? Maybe just make the `EmptyBoundsSpec` var a pointer type itself?

adamperlin commented:
@nathanielc Addressed your review comments. I also added a commit removing flux functions' dependence on the `(*Administration.ResolveTime)` function because it seemed confusing to do bounds resolution both there and in the planner with the same end result.

nathanielc commented on query/execute/executor.go:
```diff
@@ -128,13 +128,24 @@ func (es *executionState) createNode(ctx context.Context, pr *plan.Procedure, no
+			}
+		} else {
+			// If bounds are empty, standardize them to the default EmptyBounds
+			streamContext.bounds = EmptyBounds
```
Why is this needed? All consumers should be using IsEmpty to check if bounds are empty instead of comparing against the EmptyBounds object.

nathanielc commented on query/functions/window.go:
```diff
@@ -185,10 +185,17 @@ func createWindowTransformation(id execute.DatasetID, mode execute.AccumulationM
+	// TODO(adamperlin): could also use empty bounds by default in this case
+	if bounds == nil {
+		return nil, nil, errors.New("nil bounds passed to window")
+	}
```
EmptyBounds are different than nil bounds. The current behavior of returning an error is correct. We will need to validate that all `window` procedures have non nil bounds, but that change can be part of the empty windows change since that is when it matters. 

Can you remove the TODO? This code is correct as is.

adamperlin commented:
@nathanielc Should be ready for a final look. I changed the IgnoreBounds parameter of the window function as part of this PR, but I can move those changes over to another branch if you'd like. I'll resolve the merge conflicts if those changes stay part of this PR.

nathanielc commented on query/bounds.go:
```diff
@@ -0,0 +1,24 @@
+	Stop  Time
+}
+
+var EmptyBounds = Bounds{
```
As far as I can tell this is only used in tests. If so lets define it in the tests package.

nathanielc commented on query/execute/bounds.go:
```diff
@@ -26,6 +26,11 @@ var AllTime = Bounds{
 	Stop:  MaxTime,
 }
 
+var EmptyBounds = &Bounds{
```
Same here.
